src/V3EmitCImp.cpp

```diff
@@ -19,10 +19,138 @@
 
 #include "V3Global.h"
 #include "V3EmitC.h"
+#include "V3Ast.h"
 #include "V3EmitCFunc.h"
+#include "V3String.h"
+#include "V3UniqueNameMap.h"
 
+#include <map>
+#include <set>
 #include <vector>
-#include <unordered_set>
+
+//######################################################################
+// Utility functions
+
+static const AstNodeModule* getModp(const AstNode* nodep) {
+    const AstNodeModule* const modp = VN_CAST_CONST(nodep->user4p(), NodeModule);
+    UASSERT_OBJ(modp, nodep, "Missing module pointer " << nodep);
+    return modp;
+}
+
+//######################################################################
+// Visitor that gathers the headers required by an AstCFunc
+
+class EmitCGatherDependencies final : AstNVisitor {
+    // Ordered set, as it is used as a key in another map.
+    std::set<string> m_dependencies;
+
+    // METHODS
+    void addSymsDependency() { m_dependencies.insert(EmitCBaseVisitor::symClassName()); }
+    void addModDependency(const AstNodeModule* modp) {
+        if (const AstClass* const classp = VN_CAST_CONST(modp, Class)) {
+            m_dependencies.insert(EmitCBaseVisitor::prefixNameProtect(classp->classOrPackagep()));
+        } else {
+            m_dependencies.insert(EmitCBaseVisitor::prefixNameProtect(modp));
+        }
+    }
+    void addDTypeDependency(const AstNodeDType* nodep) {
+        if (const AstClassRefDType* const dtypep = VN_CAST_CONST(nodep, ClassRefDType)) {
+            m_dependencies.insert(
+                EmitCBaseVisitor::prefixNameProtect(dtypep->classp()->classOrPackagep()));
+        }
+    }
+    void addSelfDependency(const string& selfPointer, AstNode* nodep) {
+        if (VString::startsWith(selfPointer, "this")) {
```

Comment:
Will this break with a public or something named "thisthing"?

Reply:
This is fine , because this only checks the selfPointer set in V3Descope, which has only very specific forms (basically `this`, `this->.*`, or `(&vlSymsp->.*`